### PR TITLE
[RFC] Add new repair job for Git refs/heads using the reflog

### PR DESCRIPTION
**Pull request by mjcheetham**
Some users experence torn writes by Git after a crash to their branch refs (.git/refs/heads/**/*) leaving them unable to interact with their branches.

We're already fixing the user's HEAD ref if that becomes corrupt by reading the last location from the relevant reflog. This change generalises the fix to enable fixing of any ref using the corresponding reflog.

**Changes:**
- Generalise the work done by `GitHeadRepairJob` into an abstract job `GitRefsRepairJob`.
- Refactor the exising `GitHeadRepairJob` job to extend the abstract job.
- Introduce a new `GitHeadsRepairJob` that attempts to fix refs/heads/**/*.

**Questions:**
1. The only refs we'll now attempt to fix are HEAD, and refs/heads/**/*, not remotes (or tags).
Do we want to try fixing remote refs? Can't these be re-fetched from the remote? (what about tags?)

**TODO:**
- [x] ~Add functional test of the new repair job for corrupt refs/heads/**/*~ Done.

### CODE REVIEW

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        protected GitRefsRepairJob(ITracer tracer, TextWriter output, GVFSEnlistment enlistment) 
+            : base(tracer, output, enlistment)
+        {
+        }
+
+        public override IssueType HasIssue(List<string> messages)
+        {
+            int numBadRefs = 0;
+
+            foreach (string @ref in this.GetRefs())
```
Comment (kewillford) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: What is the `@` for?

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+            return numBadRefs > 0 ? IssueType.Fixable : IssueType.None;
+        }
+
+        /// <summary>
+        /// Fixes the refs using the reflog to find the last SHA.
+        /// </summary>
+        public override FixResult TryFixIssues(List<string> messages)
+        {
+            IEnumerable<string> badRefs = this.GetRefs().Where(x => !this.TryParseRef(x));
```
Comment (kewillford) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: Why did you use a `foreach` above and then use Linq here?  Can we be use the same in both?

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                refContents = File.ReadAllText(refPath).Trim();
+            }
+            catch (IOException ex)
+            {
+                messages?.Add($"IOException while reading '{refPath}': {ex.Message}");
+                return false;
+            }
+
+            if (this.IsValidRefContents(fullSymbolicRef, refContents))
```
Comment (kewillford) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: How are we resolving symbolic refs that have a symbolic ref?  For example files in `refs/heads/`  can have a ref in them and don't have to be a sha by running `git symbolic-ref refs/heads/branch2 refs/heads/branch1`

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                refLog = null;
+                error = $"IOException while reading reflog '{fullSymbolicRef}': " + ex.Message;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryParseRef(string fullSymbolicRef, List<string> messages = null)
```
Comment (kewillford) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: Rename `messages` to `errorMessages` and maybe it should just be and `out string errorMessage` since it doesn't look like we are adding multiple messages?

Comment (kewillford) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: Looks like this could be applied to multiple places in these changes.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerTestCase/RepairTests.cs [hunk 5]
```diff
@@ -49,6 +49,42 @@ public void FixesCorruptHeadSymRef()
             this.RepairWithConfirmShouldFix();
         }
 
+        [TestCase]
+        public void FixesCorruptHeadsSha()
```
Comment (kewillford) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerTestCase/RepairTests.cs]: Not sure if we have any but will you add/make sure there are some tests where the ref is good and doesn't need fixed and where the reflog file for the ref is missing.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 1 above)
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: `ref` is a C# keyword. Prefixing with @ escapes the keyword. Could alternatively pick a different name for this variable? `reference` perhaps?

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 4 above)
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: The `abstract class RepairJob` defines the `HasIssue` and `TryFixIssue` methods and gives you a `List<string>` for all messages to be output as part of the job execution (not just errors).

Keeping the list seems simpler as the caller can just pass in its reference to `messages` to all the called methods. It also avoids the caller(s) having to do a null check for the returned string, then an add to it's list if non-null.

Also I was trying to keep the original implementation of `GitHeadRepairJob` as much as possible since it was pre-existing. 😄

File: GVFS/GVFS.Common/Paths.Shared.cs [hunk 6]
```diff
@@ -68,5 +68,24 @@ public static string ConvertPathToGitFormat(string path)
...
+            // Ensure trailing slash on the rootPath
+            if (rootPath.Length > 0 && rootPath.Last() != Path.DirectorySeparatorChar)
+            {
+                rootPath += Path.DirectorySeparatorChar;
+            }
+
+            Uri rootUri = new Uri(rootPath, UriKind.Absolute);
+            Uri pathUri = new Uri(path, UriKind.Absolute);
+
+            Uri relativeUri = rootUri.MakeRelativeUri(pathUri);
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply GVFS/GVFS.Common/Paths.Shared.cs]: There is no 'make a relative path' in the BCL, which is a shame. The closest you can get without having to implement your own is using `System.Uri`. I was surprised the VFS4G codebase doesn't already have such a utility yet (unless I'm missing something?).

You're right about needing to unescape spaces from Uri.ToString however.. I'll look at replacing this and doing the relative path calc manually (with tests!).

[the remark this answers is not in the document]
Comment (mjcheetham) [reply GVFS/GVFS.Common/Paths.Shared.cs]: \* Grumble \* 😢 

Covering all the edge cases here is actually trickier that first imagined..

It also turns out there _is_ a [BCL API for getting a relative path](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getrelativepath?view=netcore-2.1#applies-to).. but only in .NET Core 2.0 onwards, not in .NET Standard or Framework. 😭 (this project is multi-targeted)

I've updated the current code (using `System.Uri`) to unescape spaces correctly and handle both Windows- and UNIX-style paths. Added tests too.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 3 above)
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: I pushed the symbolic ref check down into the base.

File: GVFS/GVFS/RepairJobs/GitHeadsRepairJob.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using GVFS.Common;
+using GVFS.Common.Tracing;
+
+namespace GVFS.RepairJobs
+{
+    public class GitHeadsRepairJob : GitRefsRepairJob
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitHeadsRepairJob.cs]: Renamed.

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs [hunk 8]
```diff
@@ -31,114 +30,49 @@ public override IssueType HasIssue(List<string> messages)
...
-                refContents = File.ReadAllText(refPath).Trim();
-            }
-            catch (IOException ex)
-            {
-                messages.Add("IOException while reading .git\\HEAD: " + ex.Message);
-                return false;
-            }
+        protected override bool IsValidRefContents(string fullSymbolicRef, string refContents)
+        {
+            Debug.Assert(fullSymbolicRef == GVFSConstants.DotGit.HeadName, "Only expecting to be called with the HEAD ref");
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: Removed.

File: GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs [hunk 9]
```diff
@@ -180,6 +180,20 @@ public void CloneAndMount(bool skipPrefetch)
...
+            result.ExitCode.ShouldEqual(0, $"Failed to create branch {branchName} at {sha}");
+        }
+
+        public string GetRevParse(string @ref)
+        {
+            ProcessResult result = GitProcess.InvokeProcess(this.RepoRoot, $"rev-parse {@ref}");
+            result.ExitCode.ShouldEqual(0, $"Failed to get commit for ref '{@ref}'");
+            string refSha = result.Output.Trim();
+            return refSha;
+        }
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs]: Moved to the test class.

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs [hunk 10]
```diff
@@ -182,5 +103,17 @@ private bool CanBeRepaired(List<string> messages)
 
             return true;
         }
+
+        private string GetHeadRefSha()
+        {
+            string headRefFilePath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Head);
+
+            var contents = File.ReadAllText(headRefFilePath);
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: Don't use `var`. Please scrub all uses of it in this PR.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 1 above)
Comment (sanoursa) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: I would go with `gitRef`, to avoid any chance of ambiguity

File: GVFS/GVFS.Common/Paths.Shared.cs [hunk 11]
```diff
@@ -68,5 +68,55 @@ public static string ConvertPathToGitFormat(string path)
...
+            {
+                throw new ArgumentException("Path must be absolute.", nameof(relativeTo));
+            }
+
+            if (!Path.IsPathRooted(path))
+            {
+                throw new ArgumentException("Path must be absolute.", nameof(path));
+            }
+
+            // Normalize paths
```
Comment (sanoursa) [inline GVFS/GVFS.Common/Paths.Shared.cs]: What do you mean by "normalize"? You already checked that they are full paths up above, so isn't `GetFullPath` going to be a noop? Also we have a method `IPlatformFileSystem.TryGetNormalizedPath`, which on Windows ends up calling `GetFinalPathNameByHandle` to normalize the path. 

I'm not really sure I understand the goal of this `GetRelativePath` method though. In general, paths are much more complicated than they appear and converting them in this way seems likely to be very error-prone.

File: GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            get { return @".git\refs\heads\**\*"; }
+        }
+
+        protected override IEnumerable<string> GetRefs()
+        {
+            string refsHeadsPath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Refs.Heads.RootFolder);
+            string dotGitPath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Root) + Path.DirectorySeparatorChar;
+
+            return Directory.EnumerateFiles(refsHeadsPath, "*", SearchOption.AllDirectories)
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs]: I think a better way to write this code would be to use `DirectoryInfo` to enumerate the children, so that you can keep everything in relative-path-space, and not have to convert from relative to full back to relative paths. As you've seen, paths are hard, and I can't say with any confidence that the current `GetRelativePath` implementation is correct, or even what it might be missing. We should avoid manipulating them to whatever extent possible.

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs [hunk 13]
```diff
@@ -182,5 +103,17 @@ private bool CanBeRepaired(List<string> messages)
...
         }
+
+        private string GetHeadRefSha()
+        {
+            string headRefFilePath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Head);
+
+            var contents = File.ReadAllText(headRefFilePath);
+            var sha = contents.Trim();
+
+            Debug.Assert(SHA1Util.IsValidShaFormat(sha), "Fix to HEAD ref should have written a valid SHA");
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: Please don't use `Debug.Assert`. If it's a bug, it's a bug even in production, so just throw an exception.

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs [hunk 14]
```diff
@@ -31,114 +30,36 @@ public override IssueType HasIssue(List<string> messages)
...
-            }
-
-            messages.Add("Invalid contents found in '" + GVFSConstants.DotGit.Head + "': " + refContents);
-            return false;
+        protected override IEnumerable<string> GetRefs()
+        {
+            return new[] { GVFSConstants.DotGit.HeadName };
         }
 
         private bool CanBeRepaired(List<string> messages)
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: Why didn't `CanBeRepaired` make it into the base class?

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs (hunk 13 above)
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: Also this condition seems invalid. The HEAD file does not have to contain a sha, it can contain the name of another ref.

I realize that you're currently calling this method after writing a sha, but only the calling code knows that, this method does not.

File: GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs [hunk 15]
```diff
@@ -19,7 +19,7 @@ public class GVFSFunctionalTestEnlistment
         private static readonly string ZeroBackgroundOperations = "Background operations: 0" + Environment.NewLine;
 
         private GVFSProcess gvfsProcess;
-        
+
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Tools/GVFSFunctionalTestEnlistment.cs]: Drop these noop changes

File: GVFS/GVFS.UnitTests/Common/PathsTests.cs [hunk 16]
```diff
@@ -31,5 +36,56 @@ public void CanConvertOSPathToGitFormat()
...
+        [TestCase(@"Z:\a\b\", @"C:\a\b\c\file.txt", @"C:\a\b\c\file.txt")]
+        [TestCase(@"C:\a\spacey dir\", @"C:\a\spacey dir\c\spacey file.txt", @"c\spacey file.txt")]
+        public void GetRelativePath_Windows(string relativeTo, string path, string expected)
+        {
+            var actual = Paths.GetRelativePath(relativeTo, path);
+
+            // If we're running on non-Windows the relative path is returned with platform-
+            // native path separators, even if a Windows-style path was passed in.
+            // Convert it back to the Windows-style separator "\".
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (sanoursa) [inline GVFS/GVFS.UnitTests/Common/PathsTests.cs]: I don't think we should make tests behave differently based on what platform they're running on. What is the actual goal here?

File: GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    public class GitRefsHeadsRepairJob : GitRefsRepairJob
+    {
+        public GitRefsHeadsRepairJob(ITracer tracer, TextWriter output, GVFSEnlistment enlistment)
+            : base(tracer, output, enlistment)
+        {
+        }
+
+        public override string Name
+        {
+            get { return @".git\refs\heads\**\*"; }
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs]: This isn't what your code is actually doing. Your code is checking for anything in `.git\refs\heads\*`

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                messages.Add($"Not all references could be fixed. Failed to fix {numFailures} references.");
+                return FixResult.Failure;
+            }
+
+            return FixResult.Success;
+        }
+
+        /// <summary>
+        /// Get the list of full symbolic references (eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master)
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: The example is somewhat confusing since none of the code inspects the `refs\remotes` folder.

File: GVFS/GVFS.UnitTests/Common/PathsTests.cs [hunk 19]
```diff
@@ -31,5 +40,44 @@ public void CanConvertOSPathToGitFormat()
...
+        }
+
+        private static void CreateEmptyFiles(string rootDirectory, IEnumerable<string> files)
+        {
+            foreach (string file in files)
+            {
+                string fullFilePath = Path.Combine(rootDirectory, file);
+
+                string parentDirectory = Path.GetDirectoryName(fullFilePath);
+                Directory.CreateDirectory(parentDirectory);
```
Comment (sanoursa) [inline GVFS/GVFS.UnitTests/Common/PathsTests.cs]: Unit tests must not do actual IO. Either mock out the file system, or drop this test and rely on functional tests.

File: GVFS/GVFS.UnitTests/Common/PathsTests.cs [hunk 20]
```diff
@@ -8,6 +12,11 @@ namespace GVFS.UnitTests.Common
     [TestFixture]
     public class PathsTests
     {
+        // Cannot use the System.IO.Path.(Alt)DirectorySeparatorChar field because
+        // this is different depending on the platform the tests are being run on.
+        private const char WindowsPathSeparator = '\\';
```
Comment (sanoursa) [inline GVFS/GVFS.UnitTests/Common/PathsTests.cs]: I don't think these are being used anymore?

File: GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs (hunk 17 above)
Comment (sanoursa) [reply GVFS/GVFS/RepairJobs/GitRefsHeadsRepairJob.cs]: I guess it depends on what wildcard syntax you're using. I read this as `.git\refs\heads\<some subfolder here>\*`. Either way, this is just a display string and intended to inform users, so lets remove the ambiguity. You can just say `Local branches` here and be totally clear about what you mean.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 18 above)
Comment (sanoursa) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: Even if it was pre-existing, I would drop the examples. It's potentially quite misleading to a future reader who may think that it would be desirable to repair remote refs, when it definitely would not be.

File: GVFS/GVFS.Common/Paths.Shared.cs [hunk 21]
```diff
@@ -68,5 +69,28 @@ public static string ConvertPathToGitFormat(string path)
         {
             return path.Replace(Path.DirectorySeparatorChar, GVFSConstants.GitPathSeparator);
         }
+
+        public static IEnumerable<string> GetFilesRecursive(string rootDirectory)
```
Comment (sanoursa) [inline GVFS/GVFS.Common/Paths.Shared.cs]: I would make this a private method within the repair class. While its behavior is general, we don't need it anywhere else and it's unlikely that we will. And as a private method in that class, we don't have to be super picky about its perf characteristics (e.g. how much overhead is the `yield return` syntax adding?), but I wouldn't this code getting called from a perf critical code path without more analysis.

Comment (sanoursa) [inline GVFS/GVFS.Common/Paths.Shared.cs]: I would rename this to `GetRecursiveRelativeFilePaths` - the fact that they're relative is quite important here, and it's returning paths, not files.

File: GVFS/GVFS.UnitTests/Common/PathsTests.cs (hunk 19 above)
Comment (sanoursa) [reply GVFS/GVFS.UnitTests/Common/PathsTests.cs]: See my other note, I would make the new relative path method private anyway. Functional test coverage should be adequate here.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+            error = null;
+            return true;
+        }
+
+        private bool TryParseRef(string fullSymbolicRef, List<string> messages = null)
+        {
+            string refPath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Root, fullSymbolicRef);
+            if (!File.Exists(refPath))
+            {
+                messages?.Add($"Could not find ref file for '{fullSymbolicRef}'");
```
Comment (sanoursa) [inline GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: I would indicate what path you were looking for here.

I just tried testing this code and my first attempt hit a failure to repair. My repro was to simply open up `.git/refs/heads/master` and delete its contents. The output tells me that it can't find the ref, but I'm not sure why not because it's there.

```
.git\refs\heads\**\*
    Could not find ref file for 'master'
    Repair failed. Run 'gvfs log C:\GVFSRepos\GVFS' for more info.
    Could not find reflog for ref 'master'
    Not all references could be fixed. Failed to fix 1 references.
```

Comment (sanoursa) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: I believe it's looking for a file named `.git/master`, rather than `.git/refs/heads/master`

File: GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs (hunk 14 above)
Comment (sanoursa) [reply GVFS/GVFS/RepairJobs/GitHeadRepairJob.cs]: What if HEAD is pointing to the ref that you're currently fixing? In effect, you'd be changing HEAD during one of those conflict states.

File: GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs (hunk 22 above)
Comment (mjcheetham) [reply GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs]: Ah yes. I've got my base path/ref name prefixes mixed up here. Will address in next push.

File: GVFS/GVFS.UnitTests/Common/PathsTests.cs (hunk 19 above)
Comment (mjcheetham) [reply GVFS/GVFS.UnitTests/Common/PathsTests.cs]: I'll drop it and rely on functional tests (given it's being pushed into the repair job as private).

**Comment by jrbriggs** [comment]
Closing this for now, since it's not actively under development
